### ISSUE TITLE
Turicreate 6.2.2 Style Transfer parameters not working?

### ISSUE THREAD

**Issue by saddif** [apple/turicreate#3190]
Hi, i  have been using the style transfer in Turicreate - mostly used v5.8 with some good results. Just tried V6.2.2 on google colabs and Macbook pro 2018, Catalina 10.15.4 (19E287), AMD 560x. i have noticed a couple of issues

1. Macbook - it can use the AMD GPU, but the results are a mess, not usable (although im impressed with the speed it completes the task)
2. Macbook - CPU - gives good results, but obviously slow
3. Google Colabs - after installing Tensorflow-GPU - it uses the GPU and results are similar to the macbook pro CPU results - but obviously faster
4. The advanced parameters is not recognised and thus no adjustments are made to the style loss- the same results are produced each time - regardless of how i change the parameters.

This is the code

!pip install turicreate
!pip install -y tensorflow
!pip install tensorflow-gpu
import turicreate as tc
tc.config.set_num_gpus(-1)
style = tc.load_images('style')
content = tc.load_images('content')
params = {
        'print_loss_breakdown': True,
        'content_loss_mult': 1.0,
        'style_loss_mult': [ 1e-2, 1e-3, 1e-3, 1e-2],
      
        'finetune_all_params': True,
       
        'vgg16_content_loss_layer': 2,
      'use_augmentation': True, #'aug_pca_noise' : 0.5,
        
        'input_shape': (256, 256),'batch_size': 1,
       'pretrained_weights' : True
    }

model = tc.style_transfer.create(style, content, max_iterations = 50000, _advanced_parameters=params)

The above code - minus the tensorflowGPU (instead i used !pip uninstall -y mxnet
!pip install mxnet-cu100==1.4.0.post0) would work fine on v5.8 - and i can fully customise the parameters.

So in summary
1. AMD GPU does not work
2. Parameters are not being utilised.

Has the parameters code changed some how?

Thanks

**Comment by TobyRoseman** [comment]
@saddif - Thanks for the great summary.

It's clear we have an issue with style transfer on macOS with GPU. Related issues: #3036, #3018, #2349, #2392.

**Comment by robertallenmartin** [comment]
![tctest](https://user-images.githubusercontent.com/4595271/82010549-7f339780-96c6-11ea-846f-6abc4daea028.png)

Hey Toby, So awesome to be able to run on the Mac GPU! But having similar experience as the related issues.

Steps to build...

python3 -m venv p3env
cd p3env/bin/
source activate

pip install notebook
pip install -U turicreate
pip uninstall -y tensor-flow
pip install tensorflow-gpu (caused errors)
pip uninstall -y tensorflow-gpu
pip install tensorflow (worked!)

jupyter notebook

(over to jupyter in the browser)
import turicreate as tc
tc.config.set_num_gpus(-1)
style = tc.load_images('/Users/computer/style/')
content = tc.load_images('/Users/computer/train/')
model = tc.style_transfer.create(style, content, max_iterations=20000)

Using 'image' in style_dataset as feature column and using 'image' in content_dataset as feature column
Using GPU ("AMD Radeon Pro Vega 20") to create model.
+--------------+--------------+--------------+
| Iteration    | Loss         | Elapsed Time |
+--------------+--------------+--------------+
| 1            | 103.05       | 821.416ms    |
| 2            | 40.5286      | 1.26s        |
| 3            | 155.076      | 1.76s        |

...

| 19961        | 5.15211      | 3h 0m        |
| 19971        | 42.4333      | 3h 0m        |
| 19981        | 2.74253      | 3h 1m        |
| 19991        | 4.95479      | 3h 1m        |
| 20000        | 0.850161     | 3h 1m        |
+--------------+--------------+--------------+

model.save('/Users/computer/turicreate/save')

Image used in example
JPEG image - 160 KB
Dimensions: 1000 × 667

Various JPEG Style images used to train
Approximately 623 KB - 1.5 MB
Dimensions: 3840 × 2160

Various JPEG Images used during training
Approximately 48 KB - 8 MB (average was 1.2 MB) 
Dimensions: 300 X 300 to 6000 × 4000 (average was 3264 × 2448)

Previously months ago with TuriCreate and MXnet @ 10,000 iterations on Linux GPU with CUDA 8 did a good job for me. Maybe we need to run more iterations for AMD on Mac. Will keep trying various things and watching for any changes to TuriCreate.

**Comment by saddif** [comment]
@TobyRoseman Thanks for the reply. 

Just wondering if you know how to fix the advanced parameters issues (i.e. style loss and other advanced parameters not being utilised to train the model) when using Nvidia GPU on Google Colabs (using tensorflow-gpu). Has this issue popped up before? 

@robertallenmartin - i have gone up to 40,000 iterations on the Mac AMD GPU - no improvement from the 10,000 iterations unfortunately.

Thanks for all the great work on this!

**Comment by saddif** [comment]
Hi, Just wondering if there has been any updates on being able to use the advanced parameters to fine tune the style transfer in V6.3?

@robertallenmartin - you posted a video (rather than a photo) which had style transfer applied to it  #3186 -just wondering what you used to make that? It looks amazing Thanks

**Comment by saddif** [comment]
@TobyRoseman 

some extra bits of info that might help - using Google Colabs

i downgraded the CUDA to V10.0
I installed CuDNN V7.4.1
then did this:
!pip uninstall -y tensorflow
!pip uninstall -y tensorflow_gpu
!pip install tensorflow-gpu==1.14.0

And when checking if GPU is working with

import tensorflow as tf
print(tf.test.is_gpu_available())
print(tf.config.experimental.list_physical_devices(device_type=None))

the output is:

True
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:XLA_CPU:0', device_type='XLA_CPU'), PhysicalDevice(name='/physical_device:XLA_GPU:0', device_type='XLA_GPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

however if i check using this command:
print(tf.config.list_physical_devices('GPU'))
i get an error AttributeError: module 'tensorflow._api.v1.config' has no attribute 'list_physical_devices'

If i then install turicreate - pip install turicreate and then rerun the GPU-checking command i get FALSE
If i reinstall tensorflow_GPU 1.14 (and uninstall tensorflow) - i then get TRUE

now when i try to run a model using GPU i get the following error

RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/turicreate/extensions.py in __run_class_function(self, fnname, args, kwargs)
    289         try:
--> 290             ret = self._tkclass.call_function(fnname, argument_dict)
    291         except RuntimeError as exc:

cy_model.pyx in turicreate._cython.cy_model.UnityModel.call_function()

cy_model.pyx in turicreate._cython.cy_model.UnityModel.call_function()

RuntimeError: An error occurred: AttributeError: module 'tensorflow._api.v1.config' has no attribute 'list_physical_devices'

Hope that can help with diagnosing the problem

**Comment by TobyRoseman** [comment]
TuriCreate doesn't work with TensorFlow 1.x. It's uses TensorFlow 2.x.

See our [instructions for Linux GPU support](https://github.com/apple/turicreate/blob/main/LinuxGPU.md). Start with a new environment and install TuriCreate. Take note which version of TensorFlow get installed. Then uninstall TensorFlow. Then install that same version of `tensorflow-gpu`. Also make sure your versions of CUDA and CuDNN are compatible with that version of tensorflow-gpu.
